### PR TITLE
src/mon/Paxos: Provide paxos with mon shutdown info

### PR DESCRIPTION
**Pull request by kamoltat**
**Problem:**
    
mon.a
mon.b
mon.c
mon.d
mon.e

ceph -a stop mon.d
ceph mon remove d

.
.

mon.d is down but actually did not get removed and monmap still did not get updated.

We shut down mon.d, this blocks mon.a Paxos:: begin (assuming Paxos is updating),
since mon.a (leader) sends out begin message (MMonPaxos::OP_BEGIN) to all peer
monitors including mon.d since mon.get_quorum() still is not updated and still contain mon.d.
Paxos will not proceed to the commit() phase since we did not get a reply
message(MMonPaxos::OP_ACCEPT) from mon.d. The lease of one of the monitors will eventually expire
and will call for an election, now, if the monmap proposal of ("mon remove", "name": "d")
comes in before the election happens, it will get queued to pending_finisher and eventually
be discarded once the election has started. Resulting in the remove command not taking effect.
 
However, if the election finishes before the monmap proposal, then we will be fine
because mon.get_quorum() will get updated and Paxos will not send
(MMonPaxos::OP_BEGIN) to mon.d, hence will not be blocked and will continue to the commit phase,
therefore, this problem is non-deterministic.
    
**Solution:**

The goal is to allow users to successfully shutdown
and remove a monitor without Paxos
rejecting the remove proposal because it needs **all** monitors
in the quorum to accept when we know there is a monitor that was purposely shut down. 
We provide Paxos with the additional information that the
monitor in the quorum might not be relevant if it is already
shut down. Once monmap is changed,  and Paxos is restarted,
the information about shut down monitor will be cleared because
we will start a new election and the quorum will be up to date with
the shutdown monitor being outside the quorum.

Fixes: https://tracker.ceph.com/issues/55695

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/messages/MMonShutdown.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
```
Comment (ljflores) [inline src/messages/MMonShutdown.h:6]: Would you put your own name and email here?

**Comment by kamoltat** [comment]
Test result for the following PRs combined:

https://github.com/ceph/ceph/pull/45511 - cephadm/services/cephadmservice: shutdown monitors before removing them
https://github.com/ceph/ceph/pull/46740 - src/mon/Paxos: Provide paxos with mon shutdown info
	cherry-picked only (5e9348bce9557410e8a9291b96ba1edc7b7f7a10) since the other commit are 	just loggings for debuggin purposes. 



30 Jobs - 29/30 passed 1 Dead Job unrelated to to Stray Daemon or anything to do with monitors.
https://pulpito.ceph.com/ksirivad-2022-06-28_12:50:56-orch:cephadm-wip-ksirivad-test-46749-and-45511-distro-default-smithi/

10 Jobs -  8/10 passed 1 Dead job , 1 Failure  unrelated Stray Daemon or anything to do with monitors
https://pulpito.ceph.com/ksirivad-2022-06-27_21:11:03-orch:cephadm-wip-ksirivad-test-46749-and-45511-distro-default-smithi/ 

10 Jobs - 10/10 passed 
https://pulpito.ceph.com/ksirivad-2022-06-27_20:01:50-orch:cephadm-wip-ksirivad-test-46749-and-45511-distro-default-smithi/

**Comment by kamoltat** [comment]
@gregsfortytwo Let me know if you think this is a good idea or not.

**Review by gregsfortytwo — CHANGES_REQUESTED** [review]
There are a couple of problems here:
1) Most straightforwardly, adding a new message requires all kinds of compatibility checks I'm not seeing here. If we run a new monitor and it gets into quorum and shuts down and sends this message to the old mons, they'll just crash.

2) At a higher level, I don't understand from the problem statement why mon removal is a special operation compared to anything else, and this PR adds a lot of special handling for it. Notifying the Paxos service about monitor removals is putting it in the wrong place — a mon that has been shut down will break the quorum and cause a new election, which is what's supposed to happen.

I don't remember if requests are supposed to get resubmitted by the monitor itself, or by the MonClient, when an election occurs. It's one of those, and I think we just need to fix whatever is preventing that if monitor removal requests aren't getting completed. (And now I'm thinking about it, it must be the MonClient doing resubmits, since a new election might mean lost monitors which received the client requests. In fact, see the MonClient::mon_commands member.)

**Comment by rzarzynski** [comment]
@kamoltat: ping.

**Pull request closed without merging**